Mango

landreman opened the pull request:
This "mango" branch of stellopt provides access to a bunch of optimization algorithms (presently ~36 algorithms) from the libraries PETSc, GSL, NLOpt, and HOPSPACK, using the MANGO interface at https://github.com/hiddenSymmetries/mango. To use the new algorithms, just set `opt_type` in stellopt's `&optimum` namelist to one of the new options in https://hiddensymmetries.github.io/mango/algorithms.html

Changes are limited to STELLOPTV2 (none in LIBSTELL etc).

The develop branch of stellopt was merged into the mango branch today, so merging mango into develop should have no conflicts. After merging develop into mango this morning, I verified that the BENCHMARKS/STELLOPT_TEST regression tests (LMDIF_TEST, GADE_TEST, IOTA_LMDIF, TOK_R0_DELTA, TOK_R0_RHO, QHS_LMDIF, BASIC, RECON_TOK) all give identical results on Draco for the mango and develop branches. I am happy to test on other systems if anyone would like.

make_draco.inc and make_macports_landreman.inc have new sections that illustrate linking to the MANGO library.

landreman committed:
Now compiles and links with mango, but there is a memory leak somewhere.

landreman committed:
Fixed problem where targets and sigmas were not loaded for mango

landreman committed:
Added mango bound constraints.

landreman committed:
Added lreset_axis_every_iteration

landreman committed:
Added another option for the MPI partition, when noptimizers is <0.

Commit:
Added makefile including mango on draco. Solved issue with new chisq_helicity with sigmas not defined until after objective function evaluated, but inelegant. This version used for first successful runs with 7dof scenario.

Commit:
Added GSL to draco makefile

Commit:
Added user_data argument to mango residual function, for compatibility with today's commits to mango.

Commit:
Updates to work with mango as of today (March 9 2020)

landreman committed:
Merged develop branch into mango branch. Now compiles and runs on Matt's macbook

Commit:
Updates for mango coupling on draco

Commit:
Merged latest updates from Lazerson branch into mango branch

landreman committed:
STELLOPTV2: Needed to replace EXTERNAL with PROCEDURE now that I am using BIND(C) for mango

landreman committed:
STELLOPTV2: MANGO now writes stellopt.extension file

Commit:
STELLOPTV2: stellopt_optimize_mango now returns nfev

Commit:
STELLOPTV2: Fixed issue that was causing segfault when stellopt exited

Commit:
Tweak to handling failures in stellopt-mango

Commit:
STELLOPTV2: Cleaned up printing to stdout for mango

Commit:
Merge branch 'develop' into mango

Commit:
Adjusted reference values and tolerance for STELLOPT_TEST/IOTA_LMDIF

landreman committed:
STELLOPTV2: Removed unnecessary print statement

landreman committed:
Merge branch 'mango' of https://github.com/PrincetonUniversity/STELLOPT into mango

lazersos commented on SHARE/make_draco.inc:
```diff
@@ -144,6 +146,39 @@
+#######################################################################
+
+#LMANGO = F
+LMANGO = T
```
This should be F for a few reasons.  But mostly because MANGO isn't maintained in a module on the cluster.  So unless you plan to maintain it for everyone, then I wouldn't put this here.  What I would suggest is that magno have a file which can be included which does most of what you do here (similar to PETSC's variables file).  It could looks something like:
```make
ifeq ($(LMANGO),T)
  include ${MAGNO_DIR}/variables
  MANGO_INC = -I$(MANGO_DIR)/include
  MANGO_LIB += $(MANGO_LIB_LINE)
endif
```
We should be trying to keep things simple.  This also means that if there are additions or changes to MAGNO this is handled on the MANGO side of the build.

lazersos reviewed (changes requested):
@landreman This looks great.  I have a comment on the make_machine.inc files.  I'll need to test this without MANGO first to make sure the old behavior is still correct.  Honestly, I don't understand the changes to chisq_helicity.  All the other chisq functions behave the same way.  Before we can merge this we have to test that the MANGO interface works with all the chisq values?  Many of which can't be called unless the equilibrium calculation is first run.  Also the sigmas and targets are know after the call to stellopt_init.f90.

Finally, I'm not sure we want to do the axis reset like this.  I think there should be three options.

1. Reset the axis to the initial value every iteration (this is not the RBC/ZBS value m=0 value).
2. Reset the axis to wout value (as requested by N. Pomphrey), this should be done in load_equil.
3. Set the axis to a best guess value (like you've done), but we need a more sophisticated algorithm.

landreman committed:
Modified build system to use MANGO's new mangoMakeVariables file

landreman commented:
@lazersos Thanks for taking a detailed look.
- I agree `LMANGO` default should be changed to F. Done.
- I like the idea of having mango generate a make variables file like PETSc. Done. This simplified the stellopt make_*.inc files a lot.
- I agree it is a good idea to have these multiple options for initializing the axis. How about I replace `lreset_axis_every_iteration` with a string that allows multiple options, e.g. 'previous','initial', etc. Can you explain your option 2 (something in load_equil) more?

Here is the story behind the edits to chisq_helicity. I was having mango handle the shifting and scaling of the terms in the objective function (by the sigma and target arrays). Mango requires the sigmas to be specified before the first evaluation of the objective function. This turned out to be harder than I expected for helicity because the sigma array for helicity depends on ixm_b and ixn_b. I could get the sigmas by calling stellopt_load_targets(). However, chisq_helicity segfaults in this case of calling stellopt_load_targets before stellopt_fcn because chisq_helicity uses the arrays bmnc_c, ixm_b, and ixn_b, which aren't allocated yet. Hence I only evaluate these arrays after testing that bmnc_b has been allocated. But that left the issue that the sigma array for helicity depends on the array of m and n values, hence the redundant calculation in lines 82-88. This is definitely an ugly hack. In retrospect, to avoid allocation issues like this, maybe I should just not use mango's ability to shift and scale the residuals (so mango thinks sigma(:)=1, targets(:)=1), letting stellopt handle the shifting and scaling. I'll see if I can get this working.

landreman committed:
STELLOPTV2: For mango, shifting values by target and scaling by sigma is now handled by stellopt rather than by mango. This eliminates the need for modifications in chisq_helicity.

landreman committed:
Implemented several option for initializing the magnetic axis shape for every evaluation of the objective function

Commit:
pull latest develop; resolve conficts; update pppl_gcc

zhucaoxiang commented:
@landreman Great job. I just test everything on the PPPL cluster. Most of the testing cases passed. How do you resolve the failure of the following one?
```
EXTENSION: stellopt.BASIC
==== Scalars ====
  PHIEDGE_equil: 0.514386   0.514386   0%
  CURTOR_equil: -174295.7864085   -174295.7864085   0%
  RBTOR_equil: 2.31287385883   2.31287385883   0%
  R0_equil: 1.577564814144   1.577564814144   0%
  VOLUME_equil: 2.978717214537   2.9787172145375   0%
  BETA_equil: 0.04247495078898   0.04247495078898   0%
  WP_equil: 192549.3826145   192549.3826145   0%
  ASPECT_equil: 4.365254725961   4.365254725961   0%
==== Vectors ====
  NELINE_equil: 4.991048247503e+19   4.991048247503e+19   0%
  FARADAY_equil: 8.43903915569e+18   8.439039155694e+18   0%
  PRESS_equil: 73076.67642488   73076.67642487   0%
  NE_equil: 1.0   1.0   0%
  TE_equil: 4561.23609115   4561.23609115   0%
  TI_equil: 4576.953953376   4576.953953376   0%
  BALLOON_grate: 0.0   0.0   0%
  BOOTSTRAP_equil: 7094384.57765   7094384.578059   0%
  NEO_equil: 0.008834582878308   0.002208645719896   29202%
  TXPORT_equil: 0.1011854859807   0.1011854860248   0%
=================
  STATUS: FAIL!!!!!
```

zhucaoxiang commented:
@landreman Also, is it possible for you to add a mango test/example case in the BENCHMARKS? It would also be perfect if we could update the documentation. Do you think it is possible to write a short tutorial on how to use mango algorithms in STELLOPT and insert it at https://princetonuniversity.github.io/STELLOPT/STELLOPT? Probably you can include the figure of the algorithm comparisons there. I will be more than happy to answer any questions regarding GitHub Pages.

landreman commented:
That BASIC NEO comparison has been failing even for the develop branch, at
least since I first started running the regression tests several weeks ago.

I will include a test case and add to the documentation. At the moment I've
addressed the issues Sam raised in this pull request, but want to do some
final tests on the new axis-initialization parts.

On Tue, Jun 30, 2020 at 5:08 PM Caoxiang Zhu <notifications@github.com>
wrote:

> @landreman <https://github.com/landreman> Also, is it possible for you to
> add a mango test/example case in the BENCHMARKS? It would also be perfect
> if we could update the documentation. Do you think it is possible to write
> a short tutorial on how to use mango algorithms in STELLOPT and insert it
> at https://princetonuniversity.github.io/STELLOPT/STELLOPT? Probably you
> can include the figure of the algorithm comparisons there. I will be more
> than happy if you have any questions regarding updating GitHub Pages.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PrincetonUniversity/STELLOPT/pull/68#issuecomment-652044609>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABH4ZA3HAZEFO7MVOWH3D4DRZJH6BANCNFSM4OGZ2CRA>
> .
>


-- 
=======================================
Dr. Matt Landreman
Associate Research Scientist
Institute for Research in Electronics & Applied Physics
University of Maryland
8223 Paint Branch Drive, College Park MD 20742, USA
(+1) 651-366-9306
mattland@umd.edu

landreman committed:
Merge branch 'mango' into mango_axisInit

Commit:
Removed diagnostic print statements for testing axis_init_option

Commit:
BENCHMARKS: Added MANGO example

landreman commented:
All of the suggestions by @lazersos last week have now been addressed. 
- The make_machine.inc files have been simplified, with make_all.inc now referring to a petsc-like variables file generated by MANGO. 
- The changes to chisq_helicity.f90 have been removed (by changing some details of the stellopt-mango communication.)
- The input parameter `lreset_axis_every_iteration` has been replaced by a string `axis_init_option` which has 5 options (`previous, mean, midpoint, vmec, input`). These include using the axis shape from the original input file, using vmec's `guess_axis.f` for setting the axis shape, and using the result from the previous vmec calculation. The default setting `axis_init_option="previous"` results in the same behavior that the STELLOPTV2 develop branch has had. For all other settings of `axis_init_option`, stellopt+mango now gives an optimization history that is independent of `NOPTIMIZERS` to all digits, as shown here by the 4 curves on top of each other for NOPTIMIZERS=8,4,2,1:
![image](https://user-images.githubusercontent.com/5229699/86247860-13749380-bb7b-11ea-883d-da8889b004aa.png)

There is also a new example input file in `BENCHMARKS/STELLOPT_TEST/MANGO`.

I verified again that with all the changes since last week, the STELLOPT_TEST benchmarks with compare*.py scripts produce identical behavior for the develop and mango branches.

@zhucaoxiang @lazersos Let me know if we can merge into develop now.

Commit:
update makefile and PPPL_make_file

zhucaoxiang reviewed (approved):
Works for me.
